internal/controllers/core/portforward/reconciler.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+type Reconciler struct {
+	// until PortForwarding behavior is moved off the PortForwardSubscriber and onto
+	// the reconciler, we set the reconciler to active/not with a flag (there should
+	// be no PortForwards on the EngineState anyway, but this is a failsafe).
+	isActive bool
+
+	kClient k8s.Client
+
+	// map of PortForward object name --> running forward(s)
+	activeForwards map[string]portForwardEntry
+}
+
+func NewReconciler(kClient k8s.Client) *Reconciler {
+	return &Reconciler{
+		kClient:        kClient,
+		activeForwards: make(map[string]portForwardEntry),
+	}
+}
+
+func (r *Reconciler) SetActive() {
+	// Should only be used for tests; should be deprecated when migration is complete
+	r.isActive = true
+}
+
+// Figure out the diff between what's in the data store and
+// what port-forwarding is currently active.
+func (r *Reconciler) diff(ctx context.Context, st store.RStore) (toStart []portForwardEntry, toShutdown []portForwardEntry) {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	statePFs := state.PortForwards
+
+	for name, existing := range r.activeForwards {
+		if _, onState := statePFs[name]; !onState {
+			// This port forward is no longer on the state, shut it down.
+			toShutdown = r.addToShutdown(toShutdown, existing)
+			continue
+		}
+	}
+
+	for name, desired := range statePFs {
+		existing, isActive := r.activeForwards[name]
+		if isActive {
+			// We're already running this PortForward -- do we need to do anything further?
+			if equality.Semantic.DeepEqual(existing.Spec, desired.Spec) && equality.Semantic.DeepEqual(existing.ObjectMeta, desired.ObjectMeta) {
+				// Nothing has changed, nothing to do
+				continue
+			}
+
+			// There's been a change to the spec for this PortForward, so tear down the old version
+			toShutdown = r.addToShutdown(toShutdown, existing)
+		}
+
+		// We're not running this PortForward(/the current version of this PortForward), so spin it up
+		entry := newEntry(ctx, desired)
+		toStart = r.addToStart(toStart, entry)
+	}
+	return toStart, toShutdown
+}
+
+func (r *Reconciler) addToStart(toStart []portForwardEntry, entry portForwardEntry) []portForwardEntry {
+	r.activeForwards[entry.Name] = entry
+	return append(toStart, entry)
+}
+
+func (r *Reconciler) addToShutdown(toShutdown []portForwardEntry, entry portForwardEntry) []portForwardEntry {
+	delete(r.activeForwards, entry.Name)
+	return append(toShutdown, entry)
+}
+
+func (r *Reconciler) OnChange(ctx context.Context, st store.RStore, summary store.ChangeSummary) {
+	if !r.isActive {
+		return
+	}
+
+	if summary.IsLogOnly() {
+		return
+	}
+
+	toStart, toShutdown := r.diff(ctx, st)
+	for _, entry := range toShutdown {
+		entry.cancel()
+	}
+
+	for _, entry := range toStart {
+		// Treat port-forwarding errors as part of the pod log
+		ctx := logger.CtxWithLogHandler(entry.ctx, PodLogActionWriter{
```

Comment:
should probably use ObjectLogHandler instead of a custom one

Reply:
lol I just lifted this straight from the original code. will do.